### DISCUSSION TITLE
Test-based suggestions for improving SDXL-based out- and inpainting in Krita AI plugin

### DISCUSSION THREAD

**Discussion by minsky91** [Acly/krita-ai-diffusion#962] [Ideas]
Since my current SD project involves an important outpainting component, I have performed a series of tests comparing Krita AI diffusion plugin (in which it is called ‘Expand’) and Forge’s outpainting performance, in terms of speed, quality of the output and available options to affect the result (‘versatility’). 

The tests were performed on a single 1536x864 AI-generated image with an accompanying mask image using the albedobaseXL_v21 checkpoint and various SDXL inpainting models, of which [Kataragi InpaintXL](https://huggingface.co/kataragi) looks most promising. In Forge, inpaint_only+lama ControlNet preprocessor in the img2img Inpaint mode was used, dpm++ 2M SDE sampler (for compatibility with the plugin), 20 steps and CFG of 7, denoise value fixed at 1. The only varying parameter used for testing was ControlNet’s Ending Control Step (ECS) value, which affected the result of  the model’s outpainting in various ways. Timings have been taken on a RTX 4070 Ti Super card with 16 GB VRAM. 

The prompts used: "theatre performance, river, Garden of Eden, exquisitely traced illustration, duotone, style of Kay Nielsen"
Negative: "artist name, artist signature, author, autograph, facsimile, letters, words, signature, signing, lettering, photo, photorealistic, bad quality, low resolution, blurry"

xinsir Promax was not included in the testing, since I couldn’t make it work in the same mode as the rest (it was very slow and produced only black images).
[ecomXL Inpaint](https://huggingface.co/alimama-creative/EcomXL_controlnet_inpaint) and [desitechInpaintv2 Dreamer](https://huggingface.co/destitech/controlnet-inpaint-dreamer-sdxl) models could produce a comparable image only at very low ECS values. 

_Timing results under Forge, per single outpainting (achieved after a fresh reboot of Forge):_

ECS values 0.51 - 1.0: generally, not useful in SDXL case.

ECS value 0.50: 
Kataragi SDXL model:  9 - 10 s. 
inpaint26.foocus: 9 s. in Forge

ECS value 0.25: 
Kataragi: 9-10s. 
inpaint26.foocus: 6 m. 18 s.

ECS value 0.04: 
Kataragi, desitechInpaintv2, EcomXL: 36 - 41 s.
inpaint26.foocus: took too long to finish the test.

_Krita AI diffusion plugin results:_ 
14-16 s. per single image expansion. 

Test image and the mask:
<img width="768" alt="Coddiwomple 06 - Leo AlbedoBaseXL Alchemy Environment + DarkArts 1536" src="https://github.com/user-attachments/assets/d9d7855e-fc48-4d71-9f32-171da1d9877e">
![fractal landscape envelope mask](https://github.com/user-attachments/assets/fe8d0e41-f83a-4690-90aa-c43f2adae47d)

Examples of outpainting:

![image (05) Kataragi ECS 0 5](https://github.com/user-attachments/assets/89ed193d-1a48-446e-8a61-60e65dd6340e)
Kataragi ECS 0.50

![image (06) Kataragi ECS 0 25](https://github.com/user-attachments/assets/fa42203f-7a54-4a1e-81a7-8bf136b59703)
Kataragi ECS 0.25

![image (3) fooocusv26 ECS 0 5](https://github.com/user-attachments/assets/305f8488-3ef1-4472-87fc-540a6a53360c)
 fooocusv26 ECS 0.50

![image (08) fooocusv26 ECS 0 25](https://github.com/user-attachments/assets/abca9b84-0562-4991-bdcf-07cfe12f8327)
fooocusv26 ECS 0.50

![image (10) desitechInpaintv2 ECS 0 01](https://github.com/user-attachments/assets/4cc6dd7a-7668-460a-a83a-cc2d1404e638)
desitechInpaintv2 ECS 0.01

![fractal landscape Albedo output 4](https://github.com/user-attachments/assets/0901dabd-24be-49af-932e-c2d75c65a522)
![fractal landscape Albedo output 3](https://github.com/user-attachments/assets/30b707f6-a820-436d-a891-7ac1a8ceb51d)
Krita AI plugin (Expand)

In my opinion, the Expand output produced by the plugin is subpar in artistic quality compared to the rest, limited in variety, while the feature lacks any option to tweak the result. (I have also tested outpainting in fooocus, that uses the same SDXL inpainting model as the plugin, and the results were even worse).
  
As for inpainting, the built-in inpainting SDXL model in the Krita plugin is already very good as it is, but still, in the future it would be nice to see support for custom (user-selected) models.

My suggestions based on the results of these tests:

1. To allow higher outpainting quality and variety of the output, add the functionality to support custom inpainting models (SDXL version and up), in addition to the built-in one based on the fooocus 2.6 patch. (This might be first introduced for outpainting only and later expanded for general inpainting.) In the UI, it can be called “Custom inpainting engine”, to avoid confusion with already supported selectable inpainting checkpoints.
2. To increase user’s control over the inpainting/outpainting output, add parameters to the UI equivalent to various options available under Forge/Automatic1111 inpainting, such as ControlNet choice of the preprocessor (inpaint only, inpaint+lama, …),  Starting/Ending Control Step. (These options, of course, can be presented in the plugin differently than in A1111/Forge and called in a more user-friendly manner, like it’s done in Leonardo.ai.)
3. Consider a way of combining outpainting operation with quality upscaling (such as SD Upscale) & refining.

**Comment by Acly** [comment]
Your use case is very special, it doesn't quite fit into the inpaint/outpaint problem space I think.

You can think of "inpaint" as using a mask which restricts where changes in the image are allowed. And "outpaint" a subset where the mask covers one or more canvas edges. This would apply to scenarios such as generating a single comic panel on a comic book page, and also scenarios such as yours (generating a frame around an image).

But this is not a problem inpaint models (or control-net) are trying to solve. They are specifically trained on content where the masked part is cut out of a larger context, and there is a smooth transition between masked/unmasked part, without seams or borders. As a result, these models will work towards connecting the existing content with whatever is being generated in the masked part with the greatest continuity.

So judging from the point of what outpaint (as a method/algorithm) is designed to do, the images generated by the plugin are the best results and the rest are strictly terrible. This is probably mostly due to parametrization though, not so much because of the models.

Judging from the point of what you're trying to achieve, the plugin's results are quite useless of course. I think you will get better results by not using any inpaint model at all in the plugin, since it's working against you (use "Generate (custom)" and untick "Seamless"). Note that using low ECS value of 0.25 in Forge also means you're hardly making use of the model.

If I'd want to add frames like this my attempt would be to outline the frame in the actual image, and keep some of it unmasked (or with a lower transparency). No inpaint model. Soft inpaint (aka. differential diffusion) could be helpful.

But if whatever workflow you found suits you go with it, sometimes the logical solution isn't the best one. I just don't think we can derive too many conclusions from this particular scenario when it comes to general inpaint model performance.

**Reply by minsky91** [reply]
Thank you for the comprehensive analysis of the case.

I have been doing outpainting already for a bigger part of the year, mostly to change the square aspect ratio of a previously generated image to a non square one, created this way literally hundreds of such face-lifted images, many of them came out great. My main tool for months had been Leonardo.ai, but nowadays, with a new GPU-equipped PC I own, it’s Forge or SwarmUI, sometimes fooocus. And the approach has always been just like you suggested, ‘to outline the frame in the actual image, and keep some of it unmasked (or with a lower transparency)’; it’s only in the example  in the github post that I used an entire image, as a shortcut. In all these tools it’s done with the Inpaint part of the UI, often offering an explicit Outpaint option as well, like in fooocus and Leonardo’s Canvas. (In the latter it’s actually a switch, see the attached screenshot. I believe Leonardo uses InvokeAI’s inpainting API for this whole thing.) Which  - obviously - is why I brought inpainting models for discussion in my post, but apparently mislaid the point. 

What also adds to the confusion, is that the best outpainting results I have been getting lately were obtained thanks to the Controlnet inpainting functionality under Forge/Auto1111 using the “inpaint_only+lama” preprocessing which works _only_ with dedicated inpaint models, the list of which for SDXL I have reviewed in the post. 

Framing images is not something I have just started doing, either.  While the plain outpainting have always been a hit-and-miss game, the results I have been getting in framing recently using Controlnet as described have almost always been breathtaking, even mind-bogglingly beautiful, something that I have never seen before. (Both with 1.5 and SDXL, but different). So no wonder that I at last decided to check out how it can be done with the Krita AI plugin and was a bit puzzled by being unable to get equivalent results right away. Now that you suggested using Generate (custom), I went to check it right away, and indeed, the output looks generally very much like what I get with Forge, but I found that I have to use low Strength values for this to happen.

The thing with ControlNet and inpainting models, it uses I guess clip vision functionality, so the outpainted result often echoes the internal elements of the base image which clip vision discovers, blended very organically within the outpainted part or the frame, yet following the prompt, which is how the magic is achieved. I suppose it's achievable in the plugin as well, but since it has no such interface, it will take a bit of experimenting.

But I still think it would be useful to have alternatives to foocus patch as the inpainting engine in the plugin. 

![Leonardo Canvas snapshot](https://github.com/user-attachments/assets/92a743b8-9b22-431b-aef3-3e549154dd57)

**Reply by Acly** [reply]
> The thing with ControlNet and inpainting models, it uses I guess clip vision functionality, so the outpainted result often echoes the internal elements of the base image which clip vision discovers

Those are completely separate. The plugin will use clip-vision/ip-adapter automatically for fill/expand _only if_ there is no prompt, and never in custom mode. You can add it manually with the "Reference" control mode.

I'm pretty sure Forge/A1111/Fooocus don't use it unless you set it explicitly, it's not linked to inpaint tab, or using an inpaint model/CN.

It sounds more like you'd need a way to adjust inpaint model/CN strength, since it seems to help your use case if it's applied for first few steps, but is generally counter productive.
